pkg/nsx/services/firewall.go

```diff
@@ -1033,13 +1034,31 @@ func (service *SecurityPolicyService) updatePeerExpressions(obj *v1alpha1.Securi
 	}
 	if peer.NamespaceSelector != nil {
 		if !mixedNsSelector {
+			if peer.NamespaceSelector.Size() == 0 {
+				// Since expressions list in model.Group must follow criteria from NSXT:
+				// 1. A non-empty expression list, must be of odd size
+				// 2. An expression list size is equal to or greater than 3
+				// 3. In a list, with indices starting from 0, all non-conjunction expressions must be at even indices
+				// Hence, add one more SegmentPort member condition to meet the criteria aforementioned
+				service.addOperatorIfNeeded(expressions, "AND")
+				clusterSegPortExpression := service.buildExpression(
+					"Condition", "SegmentPort",
+					fmt.Sprintf("%s|%s", util.TagScopeNCPCluster, service.getCluster()),
+					"Tag", "EQUALS", "EQUALS",
+				)
+				expressions.Add(clusterSegPortExpression)
+			}
+
 			tagValueExpression = nil
 			memberType = "Segment"
 			matchLabels = peer.NamespaceSelector.MatchLabels
 			matchExpressions = &peer.NamespaceSelector.MatchExpressions
 			matchLabelsCount = len(matchLabels)
 			// NamespaceSelector has one more built-in labels
 			matchLabelsCount += ClusterTagCount
+			if peer.NamespaceSelector.Size() == 0 {
+				matchLabelsCount += 1
```

Comment:
Do you think we should put `matchLabelsCount += 1` in above block(before line 1049)

Reply:
since code line 1056 will reassign matchLabelsCount value, So simple move matchLabelsCount += 1 in before line 1049 will not work, but we can create else branch to handle difference conditions here.